### ISSUE TITLE
Typing of parameters in APIs

### ISSUE THREAD

**Issue by jraman** [mwaskom/seaborn#2212]
Would you consider using [typing](https://docs.python.org/3/library/typing.html) in the API definitions please?  Not just to specify parameter typing, but also I believe it could lead to better API design.

Take `axes_style(style=None, rc=None)`, for example.  The `style` parameter currently stands for three things:
 * `None` -- which results in `matplotlib.rcParams` values (this is the default)
 * a string -- one of `{"white", "dark", "whitegrid", "darkgrid", "ticks"}`
 * a dict -- which directly specifies the style keys and values.

Given that it defaults to `None`, it could potentially expand to anything else and the code would have extra `if-else` conditions to handle them.

The typing for this function would be:

``` python
def axes_style(style: Union[None, str, dict] = None, rc: Union[None, dict] = None)
```

or

``` python
def axes_style(style: Optional[Union[str, dict]] = None, rc: Optional[dict] = None)
```

Looking at the above definitions may lead one to redesign the API as:

``` python
def axes_style(style: Optional[str] = None, rc: Optional[dict] = None)
```

Another variation:

``` python
def axes_style(style: str, rc: Optional[dict] = None)
```

where `style` is now required and an additional value -- e.g. `"matplotlib"` or `"default"` -- would be added to the list of accepted values to replace `None`.  (One could also rename the parameters as, for example, `name` and `map`, but that would be a larger API change).

The parameters are now better defined and IMHO this would be a benefit.

This applies to other functions as well.

In summary, using typing forces one to think more about the design of the API leading to better definition and better design.

**Comment by stdedos** [comment]
> I'd be generally opposed to this because the function signatures are already quite large and many parameters are very flexible in what they can take and aren't possible to statically type.

In general, it is always good to get some hints from our editors on what is considered acceptable. 🙏 
As a trivial example, passing a number where a list is required is never a happy place.

Function signatures will grow, but nothing is "too ugly" when <https://github.com/psf/black>'ed.
Of course "someone" would be able to do typing on separate files, but that's ill-advised (https://stackoverflow.com/a/47371248/2309247) or big effort in general (https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-library-stubs-or-py-typed-marker:~:text=Writing%20your%20own,compliant%20packages.)
